pull: log resolved version and digest for repo/non-OCI pulls

benoittgt opened the pull request:
**What this PR does / why we need it**:

Fixes: https://github.com/helm/helm/issues/31205

When users run `helm pull` without a version, OCI pulls already print “Pulled:” and “Digest:”. HTTP/repo pulls do not. Some providers (e.g. Bitnami) resolve --repo pulls to `oci://` but skip the registry client summary, so nothing is printed.

- Now we print a summary for `--repo` and direct HTTP pulls after download.
- Keep direct oci:// behavior unchanged (avoid duplicate output).
- Mirror OCI format: 
    ```
    Pulled: <chart>:<version>
    Digest: sha256:<digest>
    ```

For HTTP/repo, the digest is the archive SHA-256 (.tgz). For OCI, the digest is the manifest digest (unchanged).

I tested with command like:

```sh
helm pull ingress-nginx --repo=https://kubernetes.github.io/ingress-nginx --version 4.10.1
helm pull oci://registry-1.docker.io/bitnamicharts/nginx --version '20.0.7'
helm/bin/helm pull redis --repo=https://charts.bitnami.com/bitnami --version='22.0.4'
```

Overall, I think the solution a little bit _hacky_ but interested with your feedbacks.

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->


**Special notes for your reviewer**:

See inline pull request comments.

**If applicable**:
- [ ] this PR contains user facing changes (the `docs needed` label should be applied if so)
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

benoittgt committed:
pull: log resolved version and digest for repo/non-OCI pulls

When users run `helm pull` without a version, OCI pulls already print
“Pulled:” and “Digest:”. HTTP/repo pulls do not. Some providers (e.g.
Bitnami) resolve --repo pulls to `oci://` but skip the registry client
summary, so nothing is printed.

- Now we print a summary for `--repo` and direct HTTP pulls after
download.
- Keep direct oci:// behavior unchanged (avoid duplicate output).
- Mirror OCI format:
    ```
    Pulled: <chart>:<version>
    Digest: sha256:<digest>
    ```

For HTTP/repo, the digest is the archive SHA-256 (.tgz).
For OCI, the digest is the manifest digest (unchanged).

I tested with command like:

```sh
helm pull ingress-nginx --repo=https://kubernetes.github.io/ingress-nginx --version 4.10.1
helm pull oci://registry-1.docker.io/bitnamicharts/nginx --version '20.0.7'
helm/bin/helm pull redis --repo=https://charts.bitnami.com/bitnami --version='22.0.4'
```

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt commented on pkg/action/pull.go:
```diff
@@ -138,6 +139,37 @@ func (p *Pull) Run(chartRef string) (string, error) {
+	// For HTTP/repo pulls, the digest is the SHA-256 of the downloaded .tgz archive.
+	// For direct OCI pulls, the registry client already prints "Pulled:" and
+	// "Digest:" (manifest digest), so we do not print here to avoid duplicates.
+	if p.RepoURL != "" || !registry.IsOCI(downloadSourceRef) {
```
Not a big fan of this since it feels a bit hacky, but with OCI auto-resolving like Bitnami I was short on alternatives.

benoittgt replied:
Thanks @sabre1041 you are right. It is misleading. Because we use `sha256.New()` it is a [checksum](https://pkg.go.dev/crypto/sha256#New). 

Changed in https://github.com/helm/helm/pull/31206/commits/83789b112b43828e17bfbf6a9a1c964278a54f77

benoittgt commented on pkg/action/pull.go:
```diff
@@ -138,6 +139,37 @@ func (p *Pull) Run(chartRef string) (string, error) {
+			ver = base[i+1:]
+		}
+		if ver != "" {
+			fmt.Fprintf(&out, "Pulled: %s:%s\n", chart, ver)
```
Are we ok that it will differ a little bit as OCI classic output ?

benoittgt committed:
Properly match new output in pull cmd tests

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Print full url instead of chart name and version

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Output should be closer to the one for OCI registry

Plus handling of p.Version if we are not able to extract the version...
not sure is what we want..
Plus handling of tar.gz
Plus a little refactoring

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt commented on pkg/action/pull.go:
```diff
@@ -138,6 +140,34 @@ func (p *Pull) Run(chartRef string) (string, error) {
+		tag := chart
+		if ver != "" {
+			tag += ":" + ver
+		} else if p.Version != "" {
```
I am not sure we want this. Because it mean we can select a version but another one could be downloaded during resolving... Even less safe that url version extraction. I am ok to remove it or move the logic to somewhere else, but if we want to read the exact version we will have to use `chart/loader`.

benoittgt committed:
Deal with potential chart name with that start with hyphen

From copilot review:
> The condition i > 0 will fail to split chart names that start with a hyphen followed by a version (e.g., "-chart-1.0.0" would return "-chart-1.0.0", "" instead of "-chart", "1.0.0"). This should be i >= 0 to handle edge cases correctly.

https://github.com/helm/helm/pull/31206#discussion_r2311950405

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

TerryHowe reviewed (commented):
I like this idea, but I generally stick to the oci side of things

benoittgt committed:
Inform when we are no able to provide digest

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt committed:
Focus on tgz

https://github.com/helm/helm/pull/31206#discussion_r2328036710

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

benoittgt commented on pkg/action/pull.go:
```diff
@@ -138,6 +140,34 @@ func (p *Pull) Run(chartRef string) (string, error) {
+
+		if sum, err := sha256File(saved); err == nil {
+			fmt.Fprintf(&out, "Digest: sha256:%x\n", sum)
+		}
```
https://github.com/helm/helm/pull/31206/commits/df391c0324861b59317a5c610ab9687354efae9a

TerryHowe reviewed (dismissed):
/lgtm

benoittgt committed:
Avoid misleading terminology when displaying SHA-256 checksum

For HTTP/repo-based chart pulls, we calculate the SHA-256 checksum of the
downloaded .tgz archive file, not a manifest digest. This change updates
the output terminology to be more accurate:

- HTTP pulls: "Checksum: sha256:..." (file checksum)
- OCI pulls: "Digest: ..." (manifest digest, unchanged)

Signed-off-by: Benoit Tigeot <benoit.tigeot@lifen.fr>

sabre1041 reviewed (approved):
LGTM

mattfarina commented on pkg/action/pull.go:
```diff
@@ -138,6 +140,36 @@ func (p *Pull) Run(chartRef string) (string, error) {
+			fmt.Fprintf(&out, "Checksum: sha256:%x\n", sum)
+		} else {
+			fmt.Fprintf(&out, "Checksum failed: %v\n", err)
+		}
```
Internally, `DownloadTo` will sometimes have the hash of a file. The digest/hash is one of the things on a Repo index. I wonder if that should be exposed and checked against the calculated hash of the file. Then a message displayed if they are different.

Thoughts?

benoittgt replied:
Ok.. I will see how I can add a test for that but here is the output with a manually corrupted file.

```sh
$ HELM_TEST_CORRUPT=1  bin/helm pull stable/mysql

Pulled: stable/mysql
Checksum: sha256:3791aa11aed2d587c575be44fdd56895d0ecf62a84fe3a326e51ef0fc03208be
WARNING: Checksum mismatch!
  Expected: 35f232f0b4df50e85c6d65cbe7e5291b8ac8fab2c5a6afa8a0b816c7ce594390
  Got:      3791aa11aed2d587c575be44fdd56895d0ecf62a84fe3a326e51ef0fc03208be
```
